website/guides/advanced/merge-multiline-logs.md

```diff
@@ -0,0 +1,172 @@
+in the [guide on parsing CSV logs][guides.parsing-csv-logs-with-lua]. The underlying algorithm is the following
+following:
+
+1. Parse incoming log line as a CSV row.
+2. Check the number of fields in it.
+   1. If the number of fields matches the expected number of fields,
+      then the log line contains all necessary fields and can be
+      processed further.
+   2. Otherwise, store the log line in the state of the transform and
+      then, when the next event comes, merge the subsequent log line
+      with the previous ones and repeat parsing again.
+
+Such an algorithm can be implemented, for example, with the following transform config:
+
+```toml title="vector.toml"
+[transforms.lua]
+  inputs = []
+  type = "lua"
+  version = "2"
+  source = """
+    csv = require("csv") -- load the `lua-csv` module
+    expected_columns = 23 -- expected number of columns in incoming CSV lines
+    line_separator = "\\r\\n" -- note the double escaping required by the TOML format
+  """
+  hooks.process = """
+    function (event, emit)
+      merged_event = merge(event)
+      if merged_event == nil then -- a global variable containing the merged event
+        merged_event = event -- if it is empty, set it to the current event
+      else -- otherwise, concatenate the line in the stored merged event
+           -- with the next line
+        merged_event.log.message = merged_event.log.message ..
+                                  line_separator .. event.log.message
+      end
+
+      fields = csv.openstring(event.log.message):lines()() -- parse CSV
+      if #fields < expected_columns then
+        return -- not all fields are present in the merged event yet
+      end
+
+      -- do something with the array of the parsed fields
+      merged_event.log.csv_fields = fields -- for example, just store them in an
+                                           -- array field
+
+      emit(merged_event) -- emit the resulting event
+      merged_event = nil -- clear the merged event
+    end
+  """
+```
+
+<CodeExplanation>
+
+* The `source` option defines code which is executed when the transform is created.
+* The `hooks.process` option defines a function which is called for each incoming event.
+
+</CodeExplanation>
+
+## How It Works
+
+The merging process can be represented using the following diagram:
+
+<SVG src="/img/merge-transform.svg" />
+
+The `lua` transform has internal state, which can be accessed and modified from user-defined code
+using global variables. Initially, the state is empty, which corresponds to `merged_event` variable
+being set to `nil`.
+
+As events arrive to the transform, they cause the `merged_event` variable to hold an aggregated
+event, thus making the event non-empty.
+
+In the end, when the state holds enough data to extract all fields, a merged event is emitted and
+the state is emptied. Then the process repeats as new events arrive.
+
+## Safety Checks
```

Comment:
I really like this section. Nice work! This is where similar Lua scripts fall short.